src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs

```diff
@@ -0,0 +1,301 @@
+            {
+                lineNumber = 8,
+                columnNumber = 2,
+                url = "dotnet://debugger-test.dll/this-file-doesnt-exist.cs",
+            });
+
+            var bp1_res = await cli.SendCommand("Debugger.setBreakpointByUrl", bp1_req, token);
+
+            Assert.True(bp1_res.IsOk);
+            Assert.Empty(bp1_res.Value["locations"].Values<object>());
+            //Assert.Equal ((int)MonoErrorCodes.BpNotFound, bp1_res.Error ["code"]?.Value<int> ());
+        }
+
+        [Fact]
+        public async Task CreateGoodBreakpointAndHit()
+        {
+            var bp = await SetBreakpoint("dotnet://debugger-test.dll/debugger-test.cs", 10, 8);
+
+            var eval_req = JObject.FromObject(new
+            {
+                expression = "window.setTimeout(function() { invoke_add(); }, 1);",
+            });
+
+            await EvaluateAndCheck(
+                "window.setTimeout(function() { invoke_add(); }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 10, 8,
+                "IntAdd",
+                wait_for_event_fn: (pause_location) =>
+                {
+                    Assert.Equal("other", pause_location["reason"]?.Value<string>());
+                    Assert.Equal(bp.Value["breakpointId"]?.ToString(), pause_location["hitBreakpoints"]?[0]?.Value<string>());
+
+                    var top_frame = pause_location["callFrames"][0];
+                    Assert.Equal("IntAdd", top_frame["functionName"].Value<string>());
+                    Assert.Contains("debugger-test.cs", top_frame["url"].Value<string>());
+
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 8, 4, scripts, top_frame["functionLocation"]);
+
+                    //now check the scope
+                    var scope = top_frame["scopeChain"][0];
+                    Assert.Equal("local", scope["type"]);
+                    Assert.Equal("IntAdd", scope["name"]);
+
+                    Assert.Equal("object", scope["object"]["type"]);
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 8, 4, scripts, scope["startLocation"]);
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 14, 4, scripts, scope["endLocation"]);
+                    return Task.CompletedTask;
+                }
+            );
+        }
+
+        [Fact]
+        public async Task ConditionalBreakpointTrue()
+        {
+            var bp1_res = await SetBreakpoint("dotnet://debugger-test.dll/debugger-test.cs", 12, 8, condition:"c == 30");
+
+            await EvaluateAndCheck(
+                "window.setTimeout(function() { invoke_add(); }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 12, 8, "IntAdd");
+        }
+
+        [Fact]
+        public async Task ConditionalBreakpointFalse()
+        {
+            var bp1_res = await SetBreakpoint("dotnet://debugger-test.dll/debugger-test.cs", 12, 8, condition:"c == 40");
```

Comment:
Can we use `SetMonoBreakpoint`, so we don't depend on the line numbers?

Reply:
And how can I check if stops only in my breakpoint without the condition?

Reply:
I meant, add the condition stuff to SetMonoBreakpoint too, and then use it.

Reply:
My question is how can I add the "fall-through bp" without the line?

Reply:
Sorry, I meant `SetBreakpointInMethod`. So, you would do:

```csharp
await SetBreakpointInMethod("debugger-test.dll", "Math", "IntAdd", 3, condition); // offset 3
await SetBreakpointInMethod("debugger-test.dll", "Math", "IntAdd", 4); // offset 4 - unconditional

var pause_location = await insp.WaitFor(Inspector.PAUSE);
```
Now, we can check where it stopped. Though the problem then is .. how to distinguish the two locations.
I think we are not guaranteed to get the bp source location when we send the request. But in `SetBreakpointInMethod` we do have the actual location, so maybe the method can return that also? Then you can check that `pause_location` is stopped on the location from the first bp, and not the second one.

Reply:
Done.